### PR TITLE
Enhancements to OIDC configuration handling

### PR DESCRIPTION
**Pull request by richvdh**
In preparation for handling more than one OIDC provider, split everything to do with OIDC providers out into a new `OidcProviderConfig` object.

Also add some jsonschema validation for it, to give better errors if somebody misses important things.

part of #8927.

### CODE REVIEW

File: synapse/config/oidc_config.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 # Copyright 2020 Quentin Gliech
+# Copyright 2020 The Matrix.org Foundation C.I.C.
```
Comment (clokep) [inline synapse/config/oidc_config.py:3]:
```suggestion
# Copyright 2020-2021 The Matrix.org Foundation C.I.C.
```

File: synapse/config/oidc_config.py [hunk 2]
```diff
@@ -224,3 +198,154 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
...
+        skip_verification=oidc_config.get("skip_verification", False),
+        user_profile_method=oidc_config.get("user_profile_method", "auto"),
+        allow_existing_users=oidc_config.get("allow_existing_users", False),
+        user_mapping_provider_class=user_mapping_provider_class,
+        user_mapping_provider_config=user_mapping_provider_config,
+    )
+
+
+@attr.s
+class OidcProviderConfig:
```
Comment (clokep) [inline synapse/config/oidc_config.py:300]: This idly makes me wonder if there's a way to generate a JSON schema from an attrs class... Could be fun to play with at some point!

Should this be slots or frozen?

**Review by clokep — APPROVED** [review]

File: synapse/config/oidc_config.py (hunk 1 above)
Comment (richvdh) [reply synapse/config/oidc_config.py:3]: this shows how long it's been sitting in a branch in git... I've actually fixed it in later commits.

File: synapse/config/oidc_config.py (hunk 2 above)
Comment (richvdh) [reply synapse/config/oidc_config.py:300]: > Should this be slots or frozen?

hrm, probably. I'll do it in a different PR to save updating this one.

**Pull request merged**